scip_indexer/SCIPIndexer.cc

```diff
@@ -564,36 +585,60 @@ class AliasMap final {
+                    // There are 4 possibilities here.
+                    // 1. This is an undeclared field logically defined by `klass`.
+                    // 2. This is declared in one of the modules transitively included by `klass`.
+                    // 3. This is an undeclared field logically defined by one of `klass`'s ancestor classes.
+                    // 4. This is an undeclared field logically defined by one of the modules transitively included by
+                    //    `klass`.
+                    auto normalizedKlass = FieldResolver::normalizeParentForClassVar(gs, klass.asClassOrModuleRef(),
+                                                                                     instr->name.shortName(gs));
+                    auto namedSymRef = GenericSymbolRef::field(normalizedKlass, instr->name, bind.bind.type);
+                    if (!relMap.contains(namedSymRef.withoutType())) {
+                        auto result = fieldResolver.findUnresolvedFieldTransitive(
+                            ctx, {ctx.file, klass.asClassOrModuleRef(), instr->name}, ctx.locAt(bind.loc));
+                        ENFORCE(result.inherited.exists(),
+                                "Returned non-existent class from findUnresolvedFieldTransitive with start={}, "
+                                "field={}, file={}, loc={}",
+                                klass.exists() ? klass.toStringFullName(gs) : "<non-existent>",
+                                instr->name.exists() ? instr->name.toString(gs) : "<non-existent>",
+                                ctx.file.data(gs).path(), ctx.locAt(bind.loc).showRawLineColumn(gs))
+                        relMap.insert({namedSymRef.withoutType(), result});
                     }
+                    // no trim(...) because undeclared fields shouldn't have ::
+                    ENFORCE(trim(bind.loc) == bind.loc);
+                    this->map.insert({bind.bind.variable, {namedSymRef, bind.loc, false}});
                     continue;
                 }
                 if (sym.isFieldOrStaticField()) {
+                    // There are 3 possibilities here.
+                    // 1. This is a reference to a non-instance non-class variable.
+                    // 2. This is a reference to an instance or class variable declared by `klass`.
+                    // 3. This is a reference to an instance or class variable declared by one of `klass`'s ancestor
+                    //    classes.
+                    //
+                    // For case 3, we want to emit a scip::Symbol that uses `klass`, not the ancestor.
                     ENFORCE(!bind.loc.empty());
-                    this->map.insert(
-                        {bind.bind.variable,
-                         {GenericSymbolRef::declaredField(instr->what, bind.bind.type), trim(bind.loc), false}});
+                    auto name = instr->what.name(gs);
+                    std::string_view nameText = name.shortName(gs);
```

Comment:
Do we need extra handling for globals here? I think it should be possible to have declared globals, for which we have handling in the undeclared case.

Reply:
Seems to work fine for globals as is.